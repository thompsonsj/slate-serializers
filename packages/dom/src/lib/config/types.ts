import { ChildNode, Element } from 'domhandler'

export type MarkTransform = ({ node, attribs }: {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  node?: any;
  attribs?: { [key: string]: string }
}) => Element | undefined

interface MarkTransforms {
  [key: string]: MarkTransform
}

export type ElementTransform = ({
  node,
  attribs,
  children,
}: {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  node?: any
  attribs?: { [key: string]: string }
  children?: ChildNode[]
}) => Element | undefined

interface ElementTransforms {
  [key: string]: ElementTransform
}

export interface Config {
  markMap: { [key: string]: string[] }
  elementMap: { [key: string]: string }
  elementStyleMap?: {
    [key: string]: string
  }
  markStyleMap?: {
    [key: string]: string
  }
  markTransforms?: MarkTransforms
  elementTransforms: ElementTransforms
  defaultTag?: string
  encodeEntities?: boolean
  alwaysEncodeBreakingEntities?: boolean
  alwaysEncodeCodeEntities?: boolean
  convertLineBreakToBr?: boolean
}

type UpdaterFunction = (el: Element) => Element | string

export type HtmlUpdaterFunctionMap = Record<string, UpdaterFunction>
