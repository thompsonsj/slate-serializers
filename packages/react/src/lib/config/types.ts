import { ReactNode } from 'react'

interface MarkTagTransform {
  [key: string]: ({ node, attribs }: { node?: any; attribs?: { [key: string]: string } }) => ReactNode
}

interface ElementTagTransform {
  [key: string]: ({
    node,
    attribs,
    children,
  }: {
    node?: any
    attribs?: { [key: string]: string }
    children?: ReactNode
  }) => ReactNode
}

export interface Config {
  markTransforms?: MarkTagTransform
  elementTransforms: ElementTagTransform
}
