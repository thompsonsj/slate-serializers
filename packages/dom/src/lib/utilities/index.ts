export const hasLineBreak = (str: string) => str.match(/[\r\n]+/) !== null

export const prependSpace = (str: string) => str && ` ${str.trim()}`

const camelize = (str: string) => {
  return str.replace(/(?:^|[-])(\w)/g, (a, c) => {
    c = a.substring(0, 1) === '-' ? c.toUpperCase() : c
    return c ? c : ''
  })
}

export const parseStyleCssText = (value: string): { [key: string]: string } => {
  const output: { [key: string]: string } = {}

  if (!value) {
    return output
  }

  const style = value.split(';')

  for (const s of style) {
    const rule = s.trim()

    if (rule) {
      const ruleParts = rule.split(':')
      const key = camelize(ruleParts[0].trim())
      output[key] = ruleParts[1].trim()
    }
  }

  return output
}

export const styleToString = (style: { [key: string]: string }) => {
  return Object.keys(style).reduce(
    (acc, key) =>
      acc +
      key
        .split(/(?=[A-Z])/)
        .join('-')
        .toLowerCase() +
      ':' +
      style[key] +
      ';',
    '',
  )
}

export const isEmptyObject = (obj: any) =>
  obj && Object.keys(obj).length === 0 && Object.getPrototypeOf(obj) === Object.prototype

export const removeEmpty = (obj: {}): {} => {
  return Object.fromEntries(Object.entries(obj).filter(([_, v]) => v != null))
}

/**
 *
 * @param obj an object of any dimension
 * @param args property list to check
 * @returns undefined or property value
 */
export const getNested = (obj: any, ...args: string[]) => {
  return args.reduce((o, level) => o && o[level], obj)
}

export const encodeBreakingEntities = (str: string) => {
  const swapChar = (charToSwap: string) => {
    // that swaps characters to HTML entities
    switch (charToSwap) {
      case '&':
        return '&amp;'
      case '<':
        return '&lt;'
      case '>':
        return '&gt;'
      default:
        return charToSwap
    }
  }
  str = str.replace(/[&<>]/g, (match) => {
    return swapChar(match)
  })

  return str
}

export const intersection = (o1: { [key: string]: any }, o2: { [key: string]: any }) => {
  return Object.keys(o1)
    .concat(Object.keys(o2))
    .sort()
    .reduce((r: string[], a, i, aa) => {
      if (i && aa[i - 1] === a) {
        r.push(a)
      }
      return r
    }, [])
}
