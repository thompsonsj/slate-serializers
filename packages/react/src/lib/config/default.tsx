import React, { ReactNode } from 'react'
import { Config } from './types'
import { ulid } from 'ulid'
import { slateToDomConfig } from '@slate-serializers/dom'

const BlockQuote = ({ children }: { children: ReactNode }) => (
  <blockquote>
    <p>{children}</p>
  </blockquote>
)

export const config: Config = {
  dom: {
    ...slateToDomConfig,
  },
  react: {
    elementTransforms: {
      quote: ({ children }) => {
        return <BlockQuote key={ulid()}>{children}</BlockQuote>
      },
      link: ({ node, children = [] }) => {
        const attrs: any = {}
        if (node.newTab) {
          attrs.target = '_blank'
        }
        return (
          <a key={ulid()} href={node.url} {...attrs}>
            {children}
          </a>
        )
      },
    },
  },
}
