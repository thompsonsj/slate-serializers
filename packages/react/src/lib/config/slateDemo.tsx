import { Config } from './types'

export const config: Config = {
  elementTransforms: {},
}
