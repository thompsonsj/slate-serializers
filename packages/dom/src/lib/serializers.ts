import { AnyNode, Document } from 'domhandler'

import { config as defaultConfig } from './config/default'
import { Config } from './config/types'
import { convertSlate } from './utilities/convert-slate'

type SlateToDom = (node: any[], config?: Config) => AnyNode | ArrayLike<AnyNode>

export const slateToDom: SlateToDom = (node: any[], config = defaultConfig) => {
  if (!Array.isArray(node)) {
    return new Document([])
  }
  const document = node.map((n, index) =>
    convertSlate({
      node: n,
      config,
      isLastNodeInDocument: index === node.length - 1,
    }),
  )
  return document
}
