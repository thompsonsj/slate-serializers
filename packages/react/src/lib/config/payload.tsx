import React from 'react'
import { config as defaultConfig } from './default'
import { Config as SlateToReactConfig } from './types'

/**
 * Configuration for Payload CMS
 *
 * Tested for v1.1.21
 */

export const config: SlateToReactConfig = {
  ...defaultConfig,
  elementTransforms: {
    ...defaultConfig.elementTransforms,
    link: ({ node, children = [] }) => {
      const attrs: any = {}
      if (node.linkType) {
        attrs['data-link-type'] = node.linkType
      }
      if (node.newTab) {
        attrs.target = '_blank'
      }
      return (
        <a href={node.url} {...attrs}>
          {children}
        </a>
      )
    },
  },
}
