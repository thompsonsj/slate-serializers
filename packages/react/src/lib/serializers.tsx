import React, { ReactElement, JSXElementConstructor } from 'react'
import { Element, isTag, Text } from 'domhandler'
import { getName, textContent } from 'domutils'
import { ulid } from 'ulid'

import { convertSlate } from '@slate-serializers/dom'
import { config as slateToReactConfig } from './config/default'
import type { Config as SlateToReactConfig } from './config/types'

interface ISlateToReact {
  node: any[]
  config?: SlateToReactConfig
}

export const SlateToReact = ({ node, config = slateToReactConfig }: ISlateToReact) => {
  if (!Array.isArray(node)) {
    return <></>
  }
  const document = node.map((n, index) =>
    convertSlate({
      node: n,
      config: {
        ...config.dom,
        elementTransforms: {},
        markTransforms: {},
      },
      isLastNodeInDocument: index === node.length - 1,
      customElementTransforms: config.react.elementTransforms,
      transformText: (text) => transformText(text),
      transformElement: (element) => {
        return domElementToReactElement(element)
      },
      wrapChildren: (children) => children,
    }),
  )
  return document as any
}

const transformText = (text: Text | Element) => {
  if (isTag(text as Element)) {
    return React.createElement(getName(text as Element), {}, textContent(text as Element))
  }
  return <>{textContent(text as Text)}</>
}

const domElementToReactElement = (element: Element): ReactElement<any, string | JSXElementConstructor<any>> => {
  return React.createElement(
    getName(element),
    {
      /* Ensure all React elements have a unique key - is there a better way to do this? */
      key: ulid(),
      /* Provide all attributes as props */
      ...element.attribs,
      /* Convert key names for JSX compatibility */
      ...(element.attribs?.class && { className: element.attribs?.class }),
      /* Validate style (can convert to React style object using elementAttributeTransform or other transform functions, but it is still possible that a string will be passed) */
      ...(element.attribs?.style && typeof element.attribs.style === 'object' && { style: element.attribs?.style }),
    },
    element.children as any,
  )
}
