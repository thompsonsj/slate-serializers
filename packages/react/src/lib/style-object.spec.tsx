import renderer from 'react-test-renderer'
import { SlateToReact } from './serializers'
import { styleObjectFixtures } from '@slate-serializers/tests'
import { config as defaultReactConfig } from './config/default'
import { Config as SlateToReactConfig } from './config/types'
import { transformStyleObjectToString } from '@slate-serializers/utilities'

const reactConfig: SlateToReactConfig = {
  ...defaultReactConfig,
  dom: {
    ...defaultReactConfig.dom,
    elementAttributeTransform: ({ node }) => {
      const style = transformStyleObjectToString(node.style)
      return style ? { style } : undefined
    }
  }
}

describe('style attribute css transforms with postcss', () => {
  for (const fixture of styleObjectFixtures) {
    it(`${fixture.name}`, () => {
      const tree = renderer.create(<SlateToReact node={fixture.slate} config={reactConfig} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  }
})
