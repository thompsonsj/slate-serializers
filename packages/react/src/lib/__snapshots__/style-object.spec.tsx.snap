// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style attribute css transforms with postcss mark transform 1`] = `
<p>
  <strong>
    Paragraph
  </strong>
</p>
`;

exports[`style attribute css transforms with postcss mark transforms on multiple marks 1`] = `
<p>
  This is editable 
  <strong>
    rich
  </strong>
   text, 
  <i>
    much
  </i>
   better than a 
  <pre>
    &lt;textarea&gt;
  </pre>
  !
</p>
`;

exports[`style attribute css transforms with postcss p tag element transform 1`] = `
<p
  style="font-size: 96px; --text-color: #DD3A0A; @media screen { z-index: 1; color: var(--text-color) }"
>
  <strong>
    Paragraph
  </strong>
</p>
`;
